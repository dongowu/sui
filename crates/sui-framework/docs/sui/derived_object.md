---
title: Module `sui::derived_object`
---

Enables the creation of objects with deterministic addresses derived from a parent object's UID.
This module provides a way to generate objects with predictable addresses based on a parent UID
and a key, creating a namespace that ensures uniqueness for each parent-key combination,
which is usually how registries are built.

Key features:
- Deterministic address generation based on parent object UID and key
- Derived objects can exist and operate independently of their parent

The derived UIDs, once created, are independent and do not require sequencing on the parent
object. They can be used without affecting the parent. The parent only maintains a record of
which derived addresses have been claimed to prevent duplicates.


-  [Struct `Claimed`](#sui_derived_object_Claimed)
-  [Struct `DerivedObjectKey`](#sui_derived_object_DerivedObjectKey)
-  [Constants](#@Constants_0)
-  [Function `claim`](#sui_derived_object_claim)
-  [Function `restore`](#sui_derived_object_restore)
-  [Function `exists`](#sui_derived_object_exists)
-  [Function `derive_address`](#sui_derived_object_derive_address)


<pre><code><b>use</b> <a href="../std/ascii.md#std_ascii">std::ascii</a>;
<b>use</b> <a href="../std/bcs.md#std_bcs">std::bcs</a>;
<b>use</b> <a href="../std/option.md#std_option">std::option</a>;
<b>use</b> <a href="../std/string.md#std_string">std::string</a>;
<b>use</b> <a href="../std/vector.md#std_vector">std::vector</a>;
<b>use</b> <a href="../sui/address.md#sui_address">sui::address</a>;
<b>use</b> <a href="../sui/dynamic_field.md#sui_dynamic_field">sui::dynamic_field</a>;
<b>use</b> <a href="../sui/hex.md#sui_hex">sui::hex</a>;
<b>use</b> <a href="../sui/object.md#sui_object">sui::object</a>;
<b>use</b> <a href="../sui/tx_context.md#sui_tx_context">sui::tx_context</a>;
</code></pre>



<a name="sui_derived_object_Claimed"></a>

## Struct `Claimed`

Added as a DF to the parent's UID, to mark an ID as claimed.


<pre><code><b>public</b> <b>struct</b> <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a> <b>has</b> <b>copy</b>, drop, store
</code></pre>



<details>
<summary>Fields</summary>


<dl>
<dt>
<code>0: <a href="../sui/object.md#sui_object_ID">sui::object::ID</a></code>
</dt>
<dd>
</dd>
</dl>


</details>

<a name="sui_derived_object_DerivedObjectKey"></a>

## Struct `DerivedObjectKey`

An internal key to protect from generating the same UID twice (e.g. collide with DFs)


<pre><code><b>public</b> <b>struct</b> <a href="../sui/derived_object.md#sui_derived_object_DerivedObjectKey">DerivedObjectKey</a>&lt;K: <b>copy</b>, drop, store&gt; <b>has</b> <b>copy</b>, drop, store
</code></pre>



<details>
<summary>Fields</summary>


<dl>
<dt>
<code>0: K</code>
</dt>
<dd>
</dd>
</dl>


</details>

<a name="@Constants_0"></a>

## Constants


<a name="sui_derived_object_EObjectAlreadyExists"></a>

Tries to create an object twice with the same parent-key combination.


<pre><code><b>const</b> <a href="../sui/derived_object.md#sui_derived_object_EObjectAlreadyExists">EObjectAlreadyExists</a>: u64 = 0;
</code></pre>



<a name="sui_derived_object_EInvalidParent"></a>

Tries to restore an object that does not exist for the supplied parent.


<pre><code><b>const</b> <a href="../sui/derived_object.md#sui_derived_object_EInvalidParent">EInvalidParent</a>: u64 = 1;
</code></pre>



<a name="sui_derived_object_ENotSupported"></a>

Tries to use functionality that is not supported yet.


<pre><code><b>const</b> <a href="../sui/derived_object.md#sui_derived_object_ENotSupported">ENotSupported</a>: u64 = 2;
</code></pre>



<a name="sui_derived_object_claim"></a>

## Function `claim`

Claim a deterministic UID, using the parent's UID & any key.


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_claim">claim</a>&lt;K: <b>copy</b>, drop, store&gt;(parent: &<b>mut</b> <a href="../sui/object.md#sui_object_UID">sui::object::UID</a>, key: K): <a href="../sui/object.md#sui_object_UID">sui::object::UID</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_claim">claim</a>&lt;K: <b>copy</b> + drop + store&gt;(parent: &<b>mut</b> UID, key: K): UID {
    <b>let</b> addr = <a href="../sui/derived_object.md#sui_derived_object_derive_address">derive_address</a>(parent.to_inner(), key);
    <b>let</b> id = addr.to_id();
    // If the UID <b>has</b> never been claimed, we can generate it and <b>return</b> early.
    <b>if</b> (!df::exists_(parent, <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a>(id))) {
        <b>let</b> uid = <a href="../sui/object.md#sui_object_new_uid_from_hash">object::new_uid_from_hash</a>(addr);
        // We save the value <b>as</b> `Option&lt;UID&gt;` to allow us to have "<a href="../sui/derived_object.md#sui_derived_object_restore">restore</a>" functionality <b>for</b>
        // a derived UID.
        df::add&lt;_, Option&lt;UID&gt;&gt;(parent, <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a>(id), option::none());
        <b>return</b> uid
    };
    // IF the UID <b>has</b> been restored, we can re-<b>use</b> it.
    <b>let</b> existing_uid = df::borrow_mut&lt;_, Option&lt;UID&gt;&gt;(parent, <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a>(id));
    <b>assert</b>!(existing_uid.is_some(), <a href="../sui/derived_object.md#sui_derived_object_EObjectAlreadyExists">EObjectAlreadyExists</a>);
    <b>abort</b> <a href="../sui/derived_object.md#sui_derived_object_ENotSupported">ENotSupported</a>
    // TODO: Enable once id leak verifier <b>has</b> been removed
    // existing_uid.extract()
}
</code></pre>



</details>

<a name="sui_derived_object_restore"></a>

## Function `restore`

Return a <code>UID</code>, making it reclaimable in the future.
Note: This is not yet supported.
TODO: Should we make this public(package) or internal until we are indeed able to support
reclaims?


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_restore">restore</a>(parent: &<b>mut</b> <a href="../sui/object.md#sui_object_UID">sui::object::UID</a>, uid: <a href="../sui/object.md#sui_object_UID">sui::object::UID</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_restore">restore</a>(parent: &<b>mut</b> UID, uid: UID) {
    <b>let</b> id = uid.to_inner();
    <b>assert</b>!(df::exists_(parent, <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a>(id)), <a href="../sui/derived_object.md#sui_derived_object_EInvalidParent">EInvalidParent</a>);
    <b>let</b> claimed: &<b>mut</b> Option&lt;UID&gt; = df::borrow_mut(parent, <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a>(id));
    claimed.fill(uid);
    <b>abort</b> <a href="../sui/derived_object.md#sui_derived_object_ENotSupported">ENotSupported</a>
}
</code></pre>



</details>

<a name="sui_derived_object_exists"></a>

## Function `exists`

Checks if a provided <code>key</code> has been claimed in the past. This does not guarantee
that the UID is still live (it might have been deleted.)


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_exists">exists</a>&lt;K: <b>copy</b>, drop, store&gt;(parent: &<a href="../sui/object.md#sui_object_UID">sui::object::UID</a>, key: K): bool
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_exists">exists</a>&lt;K: <b>copy</b> + drop + store&gt;(parent: &UID, key: K): bool {
    <b>let</b> addr = <a href="../sui/derived_object.md#sui_derived_object_derive_address">derive_address</a>(parent.to_inner(), key);
    df::exists_(parent, <a href="../sui/derived_object.md#sui_derived_object_Claimed">Claimed</a>(addr.to_id()))
}
</code></pre>



</details>

<a name="sui_derived_object_derive_address"></a>

## Function `derive_address`

Given an ID and a Key, it calculates the derived address.


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_derive_address">derive_address</a>&lt;K: <b>copy</b>, drop, store&gt;(parent: <a href="../sui/object.md#sui_object_ID">sui::object::ID</a>, key: K): <b>address</b>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/derived_object.md#sui_derived_object_derive_address">derive_address</a>&lt;K: <b>copy</b> + drop + store&gt;(parent: ID, key: K): <b>address</b> {
    df::hash_type_and_key(parent.to_address(), <a href="../sui/derived_object.md#sui_derived_object_DerivedObjectKey">DerivedObjectKey</a>(key))
}
</code></pre>



</details>
