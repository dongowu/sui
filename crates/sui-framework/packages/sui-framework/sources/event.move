// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

/// Events module. Defines the `sui::event::emit` function which
/// creates and sends a custom MoveEvent as a part of the effects
/// certificate of the transaction.
///
/// Every MoveEvent has the following properties:
///  - sender
///  - type signature (`T`)
///  - event data (the value of `T`)
///  - timestamp (local to a node)
///  - transaction digest
///
/// Example:
/// ```
/// module my::marketplace {
///    use sui::event;
///    /* ... */
///    struct ItemPurchased has copy, drop {
///      item_id: ID, buyer: address
///    }
///    entry fun buy(/* .... */) {
///       /* ... */
///       event::emit(ItemPurchased { item_id: ..., buyer: .... })
///    }
/// }
/// ```
module sui::event;

use std::type_name;
use sui::dynamic_field;
use sui::accumulator;
use sui::hash;
use sui::bcs;

const ENotSystemAddress: u64 = 0;

/// Emit a custom Move event, sending the data offchain.
///
/// Used for creating custom indexes and tracking onchain
/// activity in a way that suits a specific application the most.
///
/// The type `T` is the main way to index the event, and can contain
/// phantom parameters, eg `emit(MyEvent<phantom T>)`.
public native fun emit<T: copy + drop>(event: T);

public use fun destroy_stream as EventStream.destroy;
public use fun destroy_cap as EventStreamCap.destroy;
public use fun emit_authenticated as EventStreamCap.emit;

#[allow(unused_field)]
public struct EventStreamHead has store {
    /// Merkle Mountain Range of all events in the stream.
    mmr: vector<vector<u8>>,
    /// Checkpoint sequence number at which the event stream was written.
    checkpoint_seq: u64,
    /// Number of events in the stream.
    num_events: u64,
}

/// Initialize an empty Merkle Mountain Range
fun init_mmr(): vector<vector<u8>> {
    vector::empty()
}

/// Add a new value to the MMR using the standard MMR append algorithm
fun add_to_mmr(new_val: vector<u8>, mmr: &mut vector<vector<u8>>) {
    let mut i = 0;
    let mut cur = new_val;
    
    while (i < vector::length(mmr)) {
        let r = vector::borrow_mut(mmr, i);
        if (vector::is_empty(r)) {
            *r = cur;
            return
        } else {
            cur = hash_two_to_one_via_bcs(*r, cur);
            *r = vector::empty();
        };
        i = i + 1;
    };

    // Vector length insufficient. Increase by 1.
    vector::push_back(mmr, cur);
}

/// Hash two values together using BCS serialization and Blake2b
fun hash_two_to_one_via_bcs(left: vector<u8>, right: vector<u8>): vector<u8> {
    let left_bytes = bcs::to_bytes(&left);
    let right_bytes = bcs::to_bytes(&right);
    let mut concatenated = left_bytes;
    vector::append(&mut concatenated, right_bytes);
    hash::blake2b256(&concatenated)
}

entry fun update_head(accumulator_root: &mut accumulator::AccumulatorRoot, stream_id: address, new_root: vector<u8>, event_count_delta: u64, checkpoint_seq: u64, ctx: &TxContext) {
    assert!(ctx.sender() == @0x0, ENotSystemAddress);

    let name = accumulator::accumulator_key<EventStreamHead>(stream_id);
    if (dynamic_field::exists_with_type<accumulator::Key<EventStreamHead>, EventStreamHead>(accumulator_root.id(), copy name)) {
        let head: &mut EventStreamHead = dynamic_field::borrow_mut<accumulator::Key<EventStreamHead>, EventStreamHead>(accumulator_root.id_mut(), name);
        add_to_mmr(new_root, &mut head.mmr);
        head.num_events = head.num_events + event_count_delta;
        head.checkpoint_seq = checkpoint_seq;
    } else {
        let mut initial_mmr = init_mmr();
        add_to_mmr(new_root, &mut initial_mmr);
        let head = EventStreamHead {
            mmr: initial_mmr,
            checkpoint_seq: checkpoint_seq,
            num_events: event_count_delta,
        };
        dynamic_field::add<accumulator::Key<EventStreamHead>, EventStreamHead>(accumulator_root.id_mut(), name, head);
    };
}

// A unique identifier for an event stream.
public struct EventStream has key, store {
    id: UID,
}

public fun new_event_stream(ctx: &mut TxContext): EventStream {
    EventStream {
        id: object::new(ctx),
    }
}

public fun destroy_stream(stream: EventStream) {
    let EventStream { id } = stream;
    id.delete();
}

/// A capability to write to a given event stream. 
public struct EventStreamCap has key, store {
    id: UID,
    stream_id: address,
}

public fun new_cap(stream: &EventStream, ctx: &mut TxContext): EventStreamCap {
    EventStreamCap {
        id: object::new(ctx),
        stream_id: stream.id.to_address(),
    }
}

public fun destroy_cap(cap: EventStreamCap) {
    let EventStreamCap { id, .. } = cap;
    id.delete();
}

public fun emit_authenticated<T: copy + drop>(cap: &EventStreamCap, event: T) {
    let accumulator_addr = accumulator::accumulator_address<EventStreamHead>(cap.stream_id);
    emit_authenticated_impl<EventStreamHead, T>(accumulator_addr, cap.stream_id, event);
}

public fun emit_authenticated_default<T: copy + drop>(event: T) {
    let stream_id = type_name::original_package_id<T>();
    let accumulator_addr = accumulator::accumulator_address<EventStreamHead>(stream_id);
    emit_authenticated_impl<EventStreamHead, T>(accumulator_addr, stream_id, event);
}

native fun emit_authenticated_impl<StreamHeadT, T: copy + drop>(
    accumulator_id: address,
    stream: address,
    event: T,
);

#[test_only]
/// Get the total number of events emitted during execution so far
public native fun num_events(): u32;

#[test_only]
/// Get all events of type `T` emitted during execution.
/// Can only be used in testing,
public native fun events_by_type<T: copy + drop>(): vector<T>;

#[test]
fun test_mmr_addition() {
    let mut mmr = init_mmr();
    let fixed_new_val = vector::tabulate!(32, |_| 2);

    // Initial MMR should be empty
    assert!(vector::all!(&mmr, |x| vector::is_empty(x)));

    // Round 1: Add first element - should be at position 0
    add_to_mmr(fixed_new_val, &mut mmr);
    assert!(vector::map_ref!(&mmr, |x| vector::is_empty(x)) == 
            vector[false]);

    // Round 2: Add second element - should trigger merge and clear position 0
    add_to_mmr(fixed_new_val, &mut mmr);
    assert!(vector::map_ref!(&mmr, |x| vector::is_empty(x)) == 
            vector[true, false]);

    // Round 3: Add third element - should place at position 0
    add_to_mmr(fixed_new_val, &mut mmr);
    assert!(vector::map_ref!(&mmr, |x| vector::is_empty(x)) == 
            vector[false, false]);

    // Round 4: Add fourth element - should trigger cascade merge to position 2
    add_to_mmr(fixed_new_val, &mut mmr);
    assert!(vector::map_ref!(&mmr, |x| vector::is_empty(x)) == 
            vector[true, true, false]);

    // Verify the final hash represents all 4 elements
    let x = hash_two_to_one_via_bcs(fixed_new_val, fixed_new_val);
    let y = hash_two_to_one_via_bcs(x, x);
    assert!(mmr[2] == y);
}

#[test]
fun test_mmr_with_different_values() {
    let mut mmr = init_mmr();
    
    // Create different values like we would get from different events  
    let val1 = vector[150, 121, 52, 32]; // Similar to what we see in test output
    let val2 = vector[42, 233, 246, 96];  // Different hash
    let val3 = vector[148, 79, 131, 129]; // Different hash  
    let val4 = vector[50, 231, 238, 28];  // Different hash

    // Verify these values are actually different
    assert!(val1 != val2);
    assert!(val2 != val3);
    assert!(val3 != val4);

    // Add them one by one and verify MMR behavior
    add_to_mmr(val1, &mut mmr);
    add_to_mmr(val2, &mut mmr);  
    // After second add, position 0 should be empty, position 1 should have merged hash
    assert!(vector::is_empty(&mmr[0]));
    assert!(!vector::is_empty(&mmr[1]));

    add_to_mmr(val3, &mut mmr);
    // Position 0 should now have val3, position 1 should still have the merged hash
    assert!(!vector::is_empty(&mmr[0]));
    assert!(!vector::is_empty(&mmr[1]));

    add_to_mmr(val4, &mut mmr);
    // Final state: positions 0,1 empty, position 2 has the final merged hash
    assert!(vector::is_empty(&mmr[0]));
    assert!(vector::is_empty(&mmr[1]));
    assert!(!vector::is_empty(&mmr[2]));
}
