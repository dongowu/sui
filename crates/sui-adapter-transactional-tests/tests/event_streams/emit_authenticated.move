// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

// Exercise test functions that emit authenticated events

//# init --addresses test=0x0 --accounts A B --simulator

//# publish

module test::event_streams {
    use sui::event;

    public struct Object has key, store {
        id: UID,
        event_stream: event::EventStream,
        event_stream_cap: event::EventStreamCap,
    }

    public struct AuthEvent has copy, drop {
        value: u64
    }

    public entry fun create(recipient: address, ctx: &mut TxContext) {
        let stream = event::new_event_stream(ctx);
        let cap = stream.new_cap(ctx);
        transfer::public_transfer(
            Object {
                id: object::new(ctx),
                event_stream: stream,
                event_stream_cap: cap,
            },
            recipient
        )
    }

    // test that reading o2 and updating o1 works
    public entry fun emit_event(o: &mut Object, value: u64) {
        // emit an event so the world can see the new value
        let ev = AuthEvent { value };
        o.event_stream_cap.emit(ev);
    }

    public entry fun delete(o: Object) {
        let Object { id, event_stream, event_stream_cap } = o;
        event_stream_cap.destroy();
        event_stream.destroy();
        object::delete(id);
    }
}

//# run test::event_streams::create --sender A --args @A

//# view-object 2,0

//# run test::event_streams::emit_event --sender A --args object(2,0) 10

//# create-checkpoint

//# view-object 4,0

// Checkpoint 2: Add second event - should trigger MMR merge at height 1
//# run test::event_streams::emit_event --sender A --args object(2,0) 20

//# create-checkpoint

//# view-object 4,0

// Checkpoint 3: Add third event - should place at MMR[0] 
//# run test::event_streams::emit_event --sender A --args object(2,0) 30

//# create-checkpoint

//# view-object 4,0

// Checkpoint 4: Add fourth event - should trigger cascade merge to MMR[2]
//# run test::event_streams::emit_event --sender A --args object(2,0) 40

//# create-checkpoint

//# view-object 4,0

//# run test::event_streams::delete --sender A --args object(2,0)



// Run the test with:
// $ cargo nextest run -p sui-adapter-transactional-tests emit_authenticated