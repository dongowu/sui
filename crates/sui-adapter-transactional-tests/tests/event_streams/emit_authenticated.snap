---
source: external-crates/move/crates/move-transactional-test-runner/src/framework.rs
---
processed 17 tasks

init:
A: object(0,0), B: object(0,1)

task 1, lines 8-49:
//# publish
created: object(1,0)
mutated: object(0,2)
gas summary: computation_cost: 1000000, storage_cost: 8420800,  storage_rebate: 0, non_refundable_storage_fee: 0

task 2, line 51:
//# run test::event_streams::create --sender A --args @A
created: object(2,0)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 0, non_refundable_storage_fee: 0

task 3, line 53:
//# view-object 2,0
Owner: Account Address ( A )
Version: 2
Contents: test::event_streams::Object {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(2,0),
        },
    },
    event_stream: sui::event::EventStream {
        id: sui::object::UID {
            id: sui::object::ID {
                bytes: _,
            },
        },
    },
    event_stream_cap: sui::event::EventStreamCap {
        id: sui::object::UID {
            id: sui::object::ID {
                bytes: _,
            },
        },
        stream_id: _,
    },
}

task 4, line 55:
//# run test::event_streams::emit_event --sender A --args object(2,0) 10
events: Event { package_id: test, transaction_module: Identifier("event_streams"), sender: A, type_: StructTag { address: test, module: Identifier("event_streams"), name: Identifier("AuthEvent"), type_params: [] }, contents: [10, 0, 0, 0, 0, 0, 0, 0] }
mutated: object(0,0), object(2,0)
accumulators_written: fake(4,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 3024648, non_refundable_storage_fee: 30552

task 5, line 57:
//# create-checkpoint
Checkpoint created: 1

task 6, lines 59-61:
//# view-object 4,0
Owner: Object ID: ( _ )
Version: 2
Contents: sui::dynamic_field::Field<sui::accumulator::Key<sui::event::EventStreamHead>, sui::event::EventStreamHead> {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(4,0),
        },
    },
    name: sui::accumulator::Key<sui::event::EventStreamHead> {
        address: _,
    },
    value: sui::event::EventStreamHead {
        mmr: vector[
            vector[
                35u8,
                54u8,
                160u8,
                83u8,
                239u8,
                56u8,
                153u8,
                197u8,
                82u8,
                253u8,
                29u8,
                79u8,
                1u8,
                203u8,
                185u8,
                209u8,
                42u8,
                123u8,
                15u8,
                149u8,
                209u8,
                62u8,
                200u8,
                79u8,
                144u8,
                115u8,
                204u8,
                243u8,
                51u8,
                114u8,
                163u8,
                40u8,
            ],
        ],
        checkpoint_seq: 1337u64,
        num_events: 1u64,
    },
}

task 7, line 62:
//# run test::event_streams::emit_event --sender A --args object(2,0) 20
events: Event { package_id: test, transaction_module: Identifier("event_streams"), sender: A, type_: StructTag { address: test, module: Identifier("event_streams"), name: Identifier("AuthEvent"), type_params: [] }, contents: [20, 0, 0, 0, 0, 0, 0, 0] }
mutated: object(0,0), object(2,0)
accumulators_written: fake(4,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 3024648, non_refundable_storage_fee: 30552

task 8, line 64:
//# create-checkpoint
Checkpoint created: 2

task 9, lines 66-68:
//# view-object 4,0
Owner: Object ID: ( _ )
Version: 3
Contents: sui::dynamic_field::Field<sui::accumulator::Key<sui::event::EventStreamHead>, sui::event::EventStreamHead> {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(4,0),
        },
    },
    name: sui::accumulator::Key<sui::event::EventStreamHead> {
        address: _,
    },
    value: sui::event::EventStreamHead {
        mmr: vector[
            vector[],
            vector[
                7u8,
                110u8,
                20u8,
                161u8,
                44u8,
                166u8,
                61u8,
                177u8,
                177u8,
                173u8,
                121u8,
                113u8,
                8u8,
                191u8,
                105u8,
                43u8,
                143u8,
                161u8,
                103u8,
                63u8,
                233u8,
                161u8,
                225u8,
                242u8,
                188u8,
                14u8,
                116u8,
                127u8,
                49u8,
                140u8,
                239u8,
                248u8,
            ],
        ],
        checkpoint_seq: 1337u64,
        num_events: 2u64,
    },
}

task 10, line 69:
//# run test::event_streams::emit_event --sender A --args object(2,0) 30
events: Event { package_id: test, transaction_module: Identifier("event_streams"), sender: A, type_: StructTag { address: test, module: Identifier("event_streams"), name: Identifier("AuthEvent"), type_params: [] }, contents: [30, 0, 0, 0, 0, 0, 0, 0] }
mutated: object(0,0), object(2,0)
accumulators_written: fake(4,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 3024648, non_refundable_storage_fee: 30552

task 11, line 71:
//# create-checkpoint
Checkpoint created: 3

task 12, lines 73-75:
//# view-object 4,0
Owner: Object ID: ( _ )
Version: 4
Contents: sui::dynamic_field::Field<sui::accumulator::Key<sui::event::EventStreamHead>, sui::event::EventStreamHead> {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(4,0),
        },
    },
    name: sui::accumulator::Key<sui::event::EventStreamHead> {
        address: _,
    },
    value: sui::event::EventStreamHead {
        mmr: vector[
            vector[
                86u8,
                86u8,
                139u8,
                226u8,
                253u8,
                124u8,
                245u8,
                168u8,
                123u8,
                166u8,
                2u8,
                201u8,
                233u8,
                179u8,
                112u8,
                133u8,
                113u8,
                199u8,
                225u8,
                130u8,
                22u8,
                89u8,
                58u8,
                13u8,
                136u8,
                115u8,
                41u8,
                101u8,
                241u8,
                199u8,
                132u8,
                187u8,
            ],
            vector[
                7u8,
                110u8,
                20u8,
                161u8,
                44u8,
                166u8,
                61u8,
                177u8,
                177u8,
                173u8,
                121u8,
                113u8,
                8u8,
                191u8,
                105u8,
                43u8,
                143u8,
                161u8,
                103u8,
                63u8,
                233u8,
                161u8,
                225u8,
                242u8,
                188u8,
                14u8,
                116u8,
                127u8,
                49u8,
                140u8,
                239u8,
                248u8,
            ],
        ],
        checkpoint_seq: 1337u64,
        num_events: 3u64,
    },
}

task 13, line 76:
//# run test::event_streams::emit_event --sender A --args object(2,0) 40
events: Event { package_id: test, transaction_module: Identifier("event_streams"), sender: A, type_: StructTag { address: test, module: Identifier("event_streams"), name: Identifier("AuthEvent"), type_params: [] }, contents: [40, 0, 0, 0, 0, 0, 0, 0] }
mutated: object(0,0), object(2,0)
accumulators_written: fake(4,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 3024648, non_refundable_storage_fee: 30552

task 14, line 78:
//# create-checkpoint
Checkpoint created: 4

task 15, line 80:
//# view-object 4,0
Owner: Object ID: ( _ )
Version: 5
Contents: sui::dynamic_field::Field<sui::accumulator::Key<sui::event::EventStreamHead>, sui::event::EventStreamHead> {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(4,0),
        },
    },
    name: sui::accumulator::Key<sui::event::EventStreamHead> {
        address: _,
    },
    value: sui::event::EventStreamHead {
        mmr: vector[
            vector[],
            vector[],
            vector[
                193u8,
                205u8,
                61u8,
                100u8,
                74u8,
                253u8,
                69u8,
                153u8,
                212u8,
                43u8,
                77u8,
                76u8,
                82u8,
                247u8,
                79u8,
                30u8,
                88u8,
                96u8,
                109u8,
                3u8,
                161u8,
                189u8,
                107u8,
                191u8,
                137u8,
                161u8,
                75u8,
                9u8,
                145u8,
                217u8,
                130u8,
                196u8,
            ],
        ],
        checkpoint_seq: 1337u64,
        num_events: 4u64,
    },
}

task 16, lines 82-87:
//# run test::event_streams::delete --sender A --args object(2,0)
mutated: object(0,0)
deleted: object(2,0)
unwrapped_then_deleted: object(_), object(_)
gas summary: computation_cost: 1000000, storage_cost: 988000,  storage_rebate: 3024648, non_refundable_storage_fee: 30552
